removed inverseKin from the foot swing planner

psiorx opened the pull request:
This removes the inverseKin call from the ZMP trajectory planner in favor of a simpler mechanism of generating swing phase knot points.  The plan generated is nearly identical to what inverseKin provides but is generated about 5 times faster.  This should help in the quest for online planning.

![swingplan](https://cloud.githubusercontent.com/assets/3849600/6756589/4eb65ca6-cf01-11e4-8d3e-83e87913e8d1.png)

RobotLocomotionBuild commented:
*** BUILD FAILED with compilation errors. ***

Site: drake004
Build Name: precompiled-pull-psiorx-jc-swingpitch-ik-into-RobotLocomotion-master
Build Server Directory: C:\\cygwin\	mp\\drake-precompiled-16415
Pull Request: https://github.com/RobotLocomotion/drake/pull/880
CDash: http://kobol.csail.mit.edu/cd/index.php?project=Drake&date=20150320

Most recent commits to drake included in this build:
```
8e9cc44 accidentally added iris to precompiled (but in the wrong way -- using add instead of submodule add -- because I forgot it wasn't there yet).  this reverts that change.
a6108ad update iris
2fce1a1 bringing submodules up to date
5b021cb Merge branch 'precompiled' of https://github.com/RobotLocomotion/drake-distro into precompiled
07cab96 update snopt
```

```
pt
-------------------------------------------
\
BUILD_PREFIX: /tmp/drake-precompiled-16415/build\
\

-- Building for: Visual Studio 10 2010
-- The C compiler identification is MSVC 16.0.40219.1
-- The CXX compiler identification is MSVC 16.0.40219.1
-- Check for working C compiler using: Visual Studio 10 2010
error opening key: Software\\Microsoft\\VisualStudio\\10.0\\vsmacros\\OtherProjects7

error opening key: Software\\Microsoft\\VisualStudio\\10.0\\vsmacros\\RecordingProject7

error opening key: Software\\Microsoft\\VisualStudio\\10.0\\vsmacros\\OtherProjects7

error opening key: Software\\Microsoft\\VisualStudio\\10.0\\vsmacros\\OtherProjects7

error opening key: Software\\Microsoft\\VisualStudio\\10.0\\vsmacros\\RecordingProject7

-- Check for working C compiler using: Visual Studio 10 2010 -- works
-- Detecting C compiler ABI info
error opening key: Software\\Microsoft\\VisualStudio\\10.0\\vsmacros\\OtherProjects7

error opening key: Software\\Microsoft\\VisualStudio\\10.0\\vsmacros\\RecordingProject7

error opening key: Software\\Microsoft\\VisualStudio\\10.0\\vsmacros\\OtherProjects7

error opening key: Software\\Microsoft\\VisualStudio\\10.0\\vsmacros\\OtherProjects7

error opening key: Software\\Microsoft\\VisualStudio\\10.0\\vsmacros\\RecordingProject7

-- Detecting C compiler ABI info - done
-- Check for working CXX compiler using: Visual Studio 10 2010
error opening key: Software\\Microsoft\\VisualStudio\\10.0\\vsmacros\\OtherProjects7

error opening key: Software\\Microsoft\\VisualStudio\\10.0\\vsmacros\\RecordingProject7

error opening key: Software\\Microsoft\\VisualStudio\\10.0\\vsmacros\\OtherProjects7

error opening key: Software\\Microsoft\\VisualStudio\\10.0\\vsmacros\\OtherProjects7

error opening key: Software\\Microsoft\\VisualStudio\\10.0\\vsmacros\\RecordingProject7

-- Check for working CXX compiler using: Visual Studio 10 2010 -- works
-- Detecting CXX compiler ABI info
error opening key: Software\\Microsoft\\VisualStudio\\10.0\\vsmacros\\OtherProjects7

error opening key: Software\\Microsoft\\VisualStudio\\10.0\\vsmacros\\RecordingProject7

error opening key: Software\\Microsoft\\VisualStudio\\10.0\\vsmacros\\OtherProjects7

error opening key: Software\\Microsoft\\VisualStudio\\10.0\\vsmacros\\OtherProjects7

error opening key: Software\\Microsoft\\VisualStudio\\10.0\\vsmacros\\RecordingProject7

-- Detecting CXX compiler ABI info - done
-- The Fortran compiler identification is unknown
CMake Error at CMakeLists.txt:11 (enable_language):
  No CMAKE_Fortran_COMPILER could be found.



-- Configuring incomplete, errors occurred!
See also \"C:/cygwin/tmp/drake-precompiled-16415/snopt/pod-build/CMakeFiles/CMakeOutput.log\".
See also \"C:/cygwin/tmp/drake-precompiled-16415/snopt/pod-build/CMakeFiles/CMakeError.log\".
Makefile:28: recipe for target 'configure' failed
make[2]: *** [configure] Error 1
Makefile:24: recipe for target 'pod-build/Makefile' failed
make[1]: *** [pod-build/Makefile] Error 2
Makefile:22: recipe for target 'all' failed
make: *** [all] Error 2
```

RobotLocomotionBuild commented:
Site: drake006
Build Name: master-pull-psiorx-jc-swingpitch-ik-into-RobotLocomotion-master
Build Server Directory: /tmp/drake-master-7839/drake
Pull Request: https://github.com/RobotLocomotion/drake/pull/880
CDash: http://kobol.csail.mit.edu/cd/viewTest.php?onlyfailed&buildid=11943

Most recent commits to drake included in this build:
```
8e3034d Merge branch 'jc-swingpitch-ik' of https://github.com/psiorx/drake
18c0f72 Merge pull request #862 from avalenzu/akv-robustify-testURDFcolllisionmex
b16d872 Merge pull request #859 from avalenzu/akv-RBMURDF-warn-once
6c1e4cc removed inverseKin from the foot swing planner
2ca047d Move warnOnce functionality into RBM
```

372 tests run:
 357 passed
 0 cancelled (with missing dependencies )
 4 timed out
 11 failed
  2 <a href=\"https://github.com/RobotLocomotion/drake/issues/18\">known issue #18</a>: Singularities in the euler-angle floating base joint cause infrequent errors on unit tests with random initial conditions
  1 <a href=\"https://github.com/RobotLocomotion/drake/issues/310\">known issue #310</a>: testLCMPlant fails on parallel build because it can't keep up with realtime
  1 <a href=\"https://github.com/RobotLocomotion/drake/issues/311\">known issue #311</a>: testFloatingBase test can fail to match tolerance
  4 <a href=\"https://github.com/RobotLocomotion/drake/issues/374\">known issue #374</a>: failing imu unit tests
  1 <a href=\"https://github.com/RobotLocomotion/drake/issues/466\">known issue #466</a>: implement Jdot*v for drakeFunction.kinematic methods
  1 <a href=\"https://github.com/RobotLocomotion/drake/issues/585\">known issue #585</a>: testIndividualCentersOfPressure fails on win64
  1 <a href=\"https://github.com/RobotLocomotion/drake/issues/822\">known issue #822</a>: testURDF fails now that it is parsing every urdf in drake
  0 unknown

examples/Airplane2D/runDircolWithObs (timed out)

examples/Atlas/test/testRunningPlanner (timed out)

examples/DubinsCar/runFunnel (timed out)

examples/Atlas/test/testSplitWalking (timed out)

RussTedrake commented:
waiting for code review from @rdeits

RobotLocomotionBuild commented:
Site: drake001.csail.mit.edu
Build Name: master-pull-psiorx-jc-swingpitch-ik-into-RobotLocomotion-master
Build Server Directory: /private/tmp/drake-master-23037/drake
Pull Request: https://github.com/RobotLocomotion/drake/pull/880
CDash: http://kobol.csail.mit.edu/cd/viewTest.php?onlyfailed&buildid=11970

Most recent commits to drake included in this build:
```
f6e40b5 Merge branch 'jc-swingpitch-ik' of https://github.com/psiorx/drake
0bef673 Merge pull request #874 from psiorx/jc-positionConstraintsmex
5305810 Merge pull request #877 from psiorx/jc-fix-ik-tests
e09dd56 Merge pull request #876 from avalenzu/upstream-hotfix-raycast-margins
9ed9c1a Merge pull request #868 from tkoolen/testForwardJacDotTimesV-tolerances
```

372 tests run:
 361 passed
 0 cancelled (with missing dependencies )
 1 timed out
 10 failed
  2 <a href=\"https://github.com/RobotLocomotion/drake/issues/18\">known issue #18</a>: Singularities in the euler-angle floating base joint cause infrequent errors on unit tests with random initial conditions
  1 <a href=\"https://github.com/RobotLocomotion/drake/issues/365\">known issue #365</a>: sampledFiniteTimeVerification does not handle infeasible programs gracefully
  4 <a href=\"https://github.com/RobotLocomotion/drake/issues/374\">known issue #374</a>: failing imu unit tests
  1 <a href=\"https://github.com/RobotLocomotion/drake/issues/466\">known issue #466</a>: implement Jdot*v for drakeFunction.kinematic methods
  1 <a href=\"https://github.com/RobotLocomotion/drake/issues/585\">known issue #585</a>: testIndividualCentersOfPressure fails on win64
  1 <a href=\"https://github.com/RobotLocomotion/drake/issues/822\">known issue #822</a>: testURDF fails now that it is parsing every urdf in drake
  0 unknown

examples/DubinsCar/runFunnel (timed out)

rdeits commented on systems/robotInterfaces/@Biped/planSwingPitched.m:
```diff
@@ -196,23 +150,13 @@ function add_foot_origin_knot(swing_pose, speed)
-               WorldPositionInFrameConstraint(biped,swing_body_index,...
-                    mean(swing_toe_points_in_foot, 2), T_toe_local_to_world, [toe_pos_in_local(1); -LATERAL_TOL; toe_ht_in_local], [toe_pos_in_local(1); LATERAL_TOL; toe_ht_in_local])};
-[pose, q_latest] = solve_for_pose(constraints, q_latest);
+pose = [swing1_origin_pose(1:3)*(1-APEX_FRACTIONS(1)) + APEX_FRACTIONS(1)*swing2_origin_pose(1:3) + [0;0;toe_ht_in_local]; quat2rpy(quat_toe_off)];
```
I think this isn't exactly right. Previously, we were interpolating the initial and final toe positions using the `APEX_FRACTIONS` to find the desired position of the toe, then solving for the pose of the origin. This is instead interpolating between the initial and final origin positions using those same fractions. That's certainly equivalent if the foot doesn't rotate, but the foot *does* rotate, so there may be some discrepancy. Specifically, if the origin of the foot link is far from the toe, this interpolation may result in the foot moving too far forward or even moving backward.

rdeits commented on systems/robotInterfaces/@Biped/planSwingPitched.m:
```diff
@@ -99,32 +87,16 @@
-
-swing1_toe_points_in_toe_local = T_toe_local_to_world \ swing1_toe_points_in_world;
-swing2_toe_points_in_toe_local = T_toe_local_to_world \ swing2_toe_points_in_world;
+terrain_pts_in_toe_local = T_toe_local_to_world \ (T_local_to_world * [terrain_pts_in_local; ones(1, size(terrain_pts_in_local,2))]);
```
Good catch...I always forget about `\`

RobotLocomotionBuild commented:
Site: drake006
Build Name: master-pull-psiorx-jc-swingpitch-ik-into-RobotLocomotion-master
Build Server Directory: /tmp/drake-master-10342/drake
Pull Request: https://github.com/RobotLocomotion/drake/pull/880
CDash: http://kobol.csail.mit.edu/cd/viewTest.php?onlyfailed&buildid=12034

Most recent commits to drake included in this build:
```
9b0bfa2 Merge branch 'jc-swingpitch-ik' of https://github.com/psiorx/drake
014cb37 now plans trajectory for the toe and backs out the pose of the origin
9b11f69 Merge pull request #879 from rdeits/rd-clear-active-set-upstream
3a1a78c Merge pull request #875 from rdeits/rd-toe-off-upstream
4d3b22c const
```

373 tests run:
 358 passed
 0 cancelled (with missing dependencies )
 5 timed out
 10 failed
  2 <a href=\"https://github.com/RobotLocomotion/drake/issues/18\">known issue #18</a>: Singularities in the euler-angle floating base joint cause infrequent errors on unit tests with random initial conditions
  1 <a href=\"https://github.com/RobotLocomotion/drake/issues/310\">known issue #310</a>: testLCMPlant fails on parallel build because it can't keep up with realtime
  4 <a href=\"https://github.com/RobotLocomotion/drake/issues/374\">known issue #374</a>: failing imu unit tests
  1 <a href=\"https://github.com/RobotLocomotion/drake/issues/466\">known issue #466</a>: implement Jdot*v for drakeFunction.kinematic methods
  1 <a href=\"https://github.com/RobotLocomotion/drake/issues/585\">known issue #585</a>: testIndividualCentersOfPressure fails on win64
  1 <a href=\"https://github.com/RobotLocomotion/drake/issues/822\">known issue #822</a>: testURDF fails now that it is parsing every urdf in drake
  0 unknown

examples/Atlas/runDRCDoorTask (timed out)

examples/Atlas/test/testRunningPlanner (timed out)

examples/DubinsCar/runFunnel (timed out)

examples/Atlas/test/testSplitWalking (timed out)

examples/Atlas/runDRCPracticeTerrain (timed out)

psiorx commented:
@rdeits Can you take another look at this, please?  It now interpolates the toe position at the apex fractions and backs out the pose of the origin at the knot points.  I've also renamed some of the transformations for clarity and removed some of the redundant ones.

RobotLocomotionBuild commented:
Site: DRAKE004
Build Name: precompiled-pull-psiorx-jc-swingpitch-ik-into-RobotLocomotion-master
Build Server Directory: C:/cygwin/tmp/drake-precompiled-4517/drake
Pull Request: https://github.com/RobotLocomotion/drake/pull/880
CDash: http://kobol.csail.mit.edu/cd/viewTest.php?onlyfailed&buildid=12061

Most recent commits to drake included in this build:
```
dd9a845 Merge branch 'jc-swingpitch-ik' of https://github.com/psiorx/drake
014cb37 now plans trajectory for the toe and backs out the pose of the origin
9b11f69 Merge pull request #879 from rdeits/rd-clear-active-set-upstream
3a1a78c Merge pull request #875 from rdeits/rd-toe-off-upstream
4d3b22c const
```

368 tests run:
 262 passed
 72 cancelled (with missing dependencies  SDPSolver:9, avl:4, bullet:11, gloptipoly3:3, gurobi:8, iris:3, lcm:26, mosek:2, octomap:1, sedumi:3, yalmip:2)
 0 timed out
 34 failed
  4 <a href=\"https://github.com/RobotLocomotion/drake/issues/374\">known issue #374</a>: failing imu unit tests
  1 <a href=\"https://github.com/RobotLocomotion/drake/issues/466\">known issue #466</a>: implement Jdot*v for drakeFunction.kinematic methods
  1 <a href=\"https://github.com/RobotLocomotion/drake/issues/468\">known issue #468</a>: non-bullet builds can throw logical error in active_collision parsing in collisionDetect
  26 <a href=\"https://github.com/RobotLocomotion/drake/issues/569\">known issue #569</a>: lcppath.mexw64 and lcppath.mexw32 fails to load on win machines
  1 <a href=\"https://github.com/RobotLocomotion/drake/issues/585\">known issue #585</a>: testIndividualCentersOfPressure fails on win64
  1 <a href=\"https://github.com/RobotLocomotion/drake/issues/866\">known issue #866</a>: testJointCalibration fails to meet tolerance (w/ probability)
  0 unknown

rdeits commented:
:+1:
